File: py/lazyid.py
import time
import random
import datetime

BASE64_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"

def _encode(bit_string):
    output = ''
    for i in range(0, len(bit_string), 6):
        chunk = bit_string[i:i+6].ljust(6, '0')
        index = int(chunk, 2)
        output += BASE64_CHARS[index]
    return output

def _decode(encoded):
    bit_string = ''
    for ch in encoded:
        if ch not in BASE64_CHARS:
            raise ValueError("Invalid character in ID")
        index = BASE64_CHARS.index(ch)
        bit_string += bin(index)[2:].zfill(6)
    return bit_string

def _get_timestamp_bits():
    ms = int(time.time() * 1000)
    return bin(ms)[2:].zfill(42)

def _generate_random_bits(length):
    bits = ''
    for _ in range((length + 7) // 8):
        byte = random.getrandbits(8)
        bits += bin(byte)[2:].zfill(8)
    return bits[:length]

def _bit_to_timestamp(bit_string):
    ts_ms = int(bit_string[:42], 2)
    seconds = ts_ms // 1000
    milliseconds = ts_ms % 1000
    dt = datetime.datetime.fromtimestamp(seconds)
    return dt.strftime('%Y-%m-%d %H:%M:%S') + '.' + str(milliseconds).zfill(3)

def lazyid(id=None):
    if id is None:
        timestamp_bits = _get_timestamp_bits()
        timestamp_encoded = _encode(timestamp_bits).rjust(7, '0')

        rand_bits = _generate_random_bits(42)
        rand_encoded = _encode(rand_bits)

        return timestamp_encoded + rand_encoded
    else:
        if len(id) != 14:
            raise ValueError("Invalid ID length")

        ts_encoded = id[:7]
        rand_encoded = id[7:]

        ts_bits = _decode(ts_encoded)
        rand_bits = _decode(rand_encoded)

        return {
            'timestamp': _bit_to_timestamp(ts_bits),
            'random_bit': rand_bits
        }

File: js/lazyid.js
(function (root, factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        module.exports = factory();
    } else if (typeof define === 'function' && define.amd) {
        define(factory);
    } else {
        root.lazyid = factory();
    }
}(typeof self !== 'undefined' ? self : this, function () {
    var base64Chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_';

    function encode(bitString) {
        var output = '';
        for (var i = 0; i < bitString.length; i += 6) {
            var chunk = bitString.substr(i, 6);
            while (chunk.length < 6) chunk += '0';
            output += base64Chars[parseInt(chunk, 2)];
        }
        return output;
    }

    function decode(encoded) {
        var bitString = '';
        for (var i = 0; i < encoded.length; i++) {
            var index = base64Chars.indexOf(encoded.charAt(i));
            if (index === -1) throw new Error('Invalid character in ID');
            bitString += index.toString(2).padStart(6, '0');
        }
        return bitString;
    }

    function getTimestampBits() {
        var now = Date.now();
        var msBits = now.toString(2);
        return msBits.padStart(42, '0');
    }

    function generateRandomBits(length) {
        var bytes = new Uint8Array(Math.ceil(length / 8));
        if (typeof crypto !== 'undefined' && crypto.getRandomValues) {
            crypto.getRandomValues(bytes);
        } else if (typeof require !== 'undefined') {
            require('crypto').randomFillSync(bytes);
        } else {
            throw new Error('Secure random not available.');
        }

        var bitString = '';
        for (var i = 0; i < bytes.length; i++) {
            bitString += bytes[i].toString(2).padStart(8, '0');
        }
        return bitString.substr(0, length);
    }

    function bitToTimestamp(bitString) {
        var ts = parseInt(bitString.substr(0, 42), 2);
        var date = new Date(ts);
        var pad = function (n, w) {
            return n.toString().padStart(w, '0');
        };
        return date.getFullYear() + '-' +
            pad(date.getMonth() + 1, 2) + '-' +
            pad(date.getDate(), 2) + ' ' +
            pad(date.getHours(), 2) + ':' +
            pad(date.getMinutes(), 2) + ':' +
            pad(date.getSeconds(), 2) + '.' +
            pad(ts % 1000, 3);
    }

    function lazyid(id) {
        if (!id) {
            var timestamp = getTimestampBits();
            var timestampEncoded = encode(timestamp).padStart(7, '0');

            var randBits = generateRandomBits(42);
            var randEncoded = encode(randBits);

            return timestampEncoded + randEncoded;
        } else {
            if (id.length !== 14) throw new Error('Invalid ID length');

            var tsEncoded = id.slice(0, 7);
            var randEncoded = id.slice(7, 14);

            var tsBits = decode(tsEncoded);
            var randBits = decode(randEncoded);

            return {
                timestamp: bitToTimestamp(tsBits),
                random_bit: randBits
            };
        }
    }

    return lazyid;
}));